runtime/globroots.c

```diff
@@ -198,70 +208,85 @@ CAMLexport void caml_remove_global_root(value *r)
   caml_delete_global_root(&caml_global_roots, r);
 }
 
+enum gc_root_class {
+  YOUNG,
+  OLD,
+  UNTRACKED
+};
+
+static enum gc_root_class classify_gc_root(value v)
+{
+  if(!Is_block(v)) return UNTRACKED;
+  if(Is_young(v)) return YOUNG;
+  if(Is_in_heap(v)) return OLD;
```

Comment:
It's a bit more low-level, but some parts of the GC do this: you could call `Classify_addr(v)` directly here (see `caml/address_class.h`).

Reply:
I still prefer this version, for two reasons:
- As you said, `Classify_addr` is lower level, and not part of the public interface of `caml/address_class.h`
- This version avoids a page table lookup when the value is unboxed or young.

Reply:
Indeed, I confused the lookup semantics of `Is_young` and `Is_in_heap_or_young`. You are correct that this version is better then.